### PR TITLE
feat: support snapshot with management api

### PR DESCRIPTION
**Pull request by umagnus**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

feat: support snapshot with management api

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
feat: support snapshot with management api
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/azureclients/fileclient/azure_fileclient.go [hunk 1]
```diff
@@ -200,3 +200,74 @@ func (c *Client) GetServiceProperties(ctx context.Context, resourceGroupName, ac
 func (c *Client) SetServiceProperties(ctx context.Context, resourceGroupName, accountName string, parameters storage.FileServiceProperties) (storage.FileServiceProperties, error) {
 	return c.fileServicesClient.SetServiceProperties(ctx, resourceGroupName, accountName, parameters)
 }
+
+// CreateSnapshot creates a snapshot
+func (c *Client) CreateSnapshot(ctx context.Context, resourceGroupName, accountName string, shareOptions *ShareOptions) error {
```
Comment (feiskyer) [inline pkg/azureclients/fileclient/azure_fileclient.go:205]: Could you add some unit tests for those new functions?

File: pkg/azureclients/fileclient/azure_fileclient.go [hunk 2]
```diff
@@ -84,11 +84,11 @@ func (c *Client) WithSubscriptionID(subscriptionID string) Interface {
 }
 
 // CreateFileShare creates a file share
-func (c *Client) CreateFileShare(ctx context.Context, resourceGroupName, accountName string, shareOptions *ShareOptions) error {
+func (c *Client) CreateFileShare(ctx context.Context, resourceGroupName, accountName string, shareOptions *ShareOptions, expand string) (storage.FileShare, error) {
```
Comment (andyzhangx) [inline pkg/azureclients/fileclient/azure_fileclient.go:87]: add comments for `expand` parameter, e.g. what's the available values

File: pkg/azureclients/fileclient/azure_fileclient.go [hunk 3]
```diff
@@ -119,14 +119,14 @@ func (c *Client) CreateFileShare(ctx context.Context, resourceGroupName, account
 	}
 	mc.Observe(rerr)
 
-	return err
+	return FileShare, err
 }
 
 // DeleteFileShare deletes a file share
-func (c *Client) DeleteFileShare(ctx context.Context, resourceGroupName, accountName, name string) error {
+func (c *Client) DeleteFileShare(ctx context.Context, resourceGroupName, accountName, name, xMsSnapshot string) error {
```
Comment (andyzhangx) [inline pkg/azureclients/fileclient/azure_fileclient.go:126]: add comments for `xMsSnapshot` parameter, e.g. what's the available values, you could copy the original comments from go sdk api

**Comment by andyzhangx** [comment]
pls also change the PR title and add release notes, it's not switch, it's support snapshot with management api

File: pkg/azureclients/fileclient/azure_fileclient.go (hunk 1 above)
Comment (umagnus) [reply pkg/azureclients/fileclient/azure_fileclient.go]: We use old function CreateFileShare to support snapshot later.

File: pkg/azureclients/fileclient/azure_fileclient.go (hunk 2 above)
Comment (umagnus) [reply pkg/azureclients/fileclient/azure_fileclient.go:89]: ok, it's added.

File: pkg/azureclients/fileclient/azure_fileclient.go (hunk 3 above)
Comment (umagnus) [reply pkg/azureclients/fileclient/azure_fileclient.go:130]: ok, it's added.

File: pkg/azureclients/fileclient/azure_fileclient.go [hunk 4]
```diff
@@ -176,10 +180,30 @@ func (c *Client) ResizeFileShare(ctx context.Context, resourceGroupName, account
...
+	}
+	mc.Observe(rerr)
+
+	return result, err
+}
+
+// ListFileShare gets a file share list
+// expand - optional, used to expand the properties within share's properties. Valid values are: deleted,
+// snapshots. Should be passed as a string with delimiter ','
+func (c *Client) ListFileShare(ctx context.Context, resourceGroupName, accountName, filter, expand string) (storage.FileShareItemsPage, error) {
```
Comment (andyzhangx) [inline pkg/azureclients/fileclient/azure_fileclient.go:203]: this func may only return part of file share snapshots, you could refer to this function to return all snapshots:
https://github.com/kubernetes-sigs/cloud-provider-azure/blob/b0d976d580d9a0252febacc7bedd9f6def79e817/pkg/azureclients/snapshotclient/azure_snapshotclient.go#L328-L342

Comment (umagnus) [reply pkg/azureclients/fileclient/azure_fileclient.go]: ok, should we use retry.Error for output or error is ok?

**Review by andyzhangx — APPROVED** [review]
/lgtm
/approve

**Pull request merged**